pkg/gadgets/run/types/factory.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"reflect"
+	"unsafe"
+
+	"golang.org/x/exp/constraints"
+)
+
+type EventFactory struct {
+	nextOffset uintptr
+	nextIndex  int
+
+	setters map[string]any
+}
+
+func NewEventFactory() *EventFactory {
+	return &EventFactory{
+		nextIndex: IndexFixed + 1,
+		setters:   make(map[string]any),
+	}
+}
+
+func (f *EventFactory) NewEvent() *Event {
+	ev := &Event{
+		Blob: make([][]byte, f.nextIndex),
+	}
+
+	ev.Blob[IndexFixed] = make([]byte, f.nextOffset)
+
+	return ev
+}
+
+type FieldType interface {
+	constraints.Integer | constraints.Float | bool
+}
+
+func FactoryAddField[T FieldType](f *EventFactory, name string) ColumnDesc {
+	offset := f.nextOffset
+
+	var zero T
+	typ := reflect.TypeOf(zero)
+
+	col := ColumnDesc{
+		Name: name,
+		Type: Type{
+			Kind: reflectTypeToKind(typ),
+		},
+		Offset:    offset,
+		BlobIndex: IndexFixed,
+	}
+
+	f.nextOffset += typ.Size()
+
+	f.setters[name] = func(ev *Event, v T) {
+		*(*T)(unsafe.Pointer(&ev.Blob[IndexFixed][offset])) = v
+	}
+
+	return col
+}
+
+func FactoryAddString(f *EventFactory, name string) ColumnDesc {
+	index := f.nextIndex
+
+	col := ColumnDesc{
+		Name: name,
+		Type: Type{
+			Kind: KindString,
+		},
+		BlobIndex: index,
+	}
+
+	f.nextIndex++
```

Comment:
These methods `FactoryAddField` and `FactoryAddString` are difficult to use because they need to be called in the exact same order between `calculateColumnsForClient` and `processEventFunc`. We have to be careful to iterate on the same object. In my previous code, I initially had problems because one was iterating over the btf fields, and the other was iterating over the types defined in the metadata yaml (and they were not in the right order).

So it makes me wonder if it's better to use hash maps keyed by the field name (`map[string][]byte`) instead of arrays (`[][]byte`).